javascript/ql/src/semmle/javascript/Closure.qll

```diff
@@ -6,101 +6,83 @@ import javascript
 
 module Closure {
   /**
-   * A call to a function in the `goog` namespace such as `goog.provide` or `goog.load`.
+   * A reference to a Closure namespace.
    */
-  class GoogFunctionCall extends CallExpr {
-    GoogFunctionCall() {
-      exists(GlobalVariable gv | gv.getName() = "goog" |
-        this.getCallee().(DotExpr).getBase() = gv.getAnAccess()
-      )
-    }
-
-    /** Gets the name of the invoked function. */
-    string getFunctionName() { result = getCallee().(DotExpr).getPropertyName() }
+  abstract class ClosureNamespaceRef extends DataFlow::Node {
+    /**
+     * Gets the namespace being referenced.
+     */
+    abstract string getClosureNamespace();
   }
 
   /**
-   * An expression statement consisting of a call to a function
-   * in the `goog` namespace.
+   * A dataflow node that returns the value of a closure namespace.
    */
-  class GoogFunctionCallStmt extends ExprStmt {
-    GoogFunctionCallStmt() { super.getExpr() instanceof GoogFunctionCall }
-
-    override GoogFunctionCall getExpr() { result = super.getExpr() }
+  abstract class ClosureNamespaceAccess extends ClosureNamespaceRef, DataFlow::SourceNode { }
 
-    /** Gets the name of the invoked function. */
-    string getFunctionName() { result = getExpr().getFunctionName() }
-
-    /** Gets the `i`th argument to the invoked function. */
-    Expr getArgument(int i) { result = getExpr().getArgument(i) }
+  /**
+   * A call to a method on the `goog.` namespace, as a closure reference.
+   */
+  abstract private class DefaultNamespaceRef extends DataFlow::MethodCallNode, ClosureNamespaceRef {
+    DefaultNamespaceRef() { this = DataFlow::globalVarRef("goog").getAMethodCall() }
 
-    /** Gets an argument to the invoked function. */
-    Expr getAnArgument() { result = getArgument(_) }
+    override string getClosureNamespace() { result = getArgument(0).asExpr().getStringValue() }
   }
 
-  abstract private class GoogNamespaceRef extends ExprOrStmt { abstract string getNamespaceId(); }
-
   /**
-   * A call to `goog.provide`.
+   * Holds if `node` is the dataflow node for a top-level expression.
```

Comment:
The term "top-level expression" deserves an explanation or example.

Reply:
I've elaborated the qldoc a bit.

Or did you mean why we bother with this predicate at all?
- that's how it worked before the PR.
- less chance of picking up random stuff in a minified file.
- more closely matches how the extractor recognizes `goog.module`.

Reply:
The expanded QLDoc is what I was after, thanks.